rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -53,10 +53,13 @@ impl OptimizerRule for ConstantFolding {
         // projected columns. With just the projected schema, it's not possible to infer types for
         // expressions that references non-projected columns within the same project plan or its
         // children plans.
+        let mut rewriter = ConstantRewriter {
```

Comment:
FYI @houqp and @Dandandan  -- as discussed on https://github.com/apache/arrow/pull/9309#discussion_r564293129 here is a proposal of how to rewrite expressions directly without quite so much copying.

It is part of my larger plan to make rewriting `LogicalPlans` easier too. 

Github kind of mangled the diff in this file, but the core change is that all code for recursing `Expr` trees that are not relevant to the constant folding is in `Expr::rewrite` now and no longer in this file